salt,charts: Grafana sidecar picks up dashboards in all namespaces

jbertran opened the pull request:
**Component**: salt,charts

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: Allowing solutions to expose dashboards that get picked up by metalk8s' Grafana deployment.


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3618

bert-e commented:
# Hello jbertran,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/3618-pick-up-dashboards-from-all-namespaces` into
`development/2.11`
* [w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast

gdemonet commented on charts/kube-prometheus-stack/charts/grafana/values.yaml:
```diff
@@ -630,12 +630,12 @@ sidecar:
     # If specified, the sidecar will look for annotation with this name to create folder and put graph here.
     # You can use this parameter together with `provider.foldersFromFilesStructure`to annotate configmaps and create folder structure.
-    folderAnnotation: null
+    folderAnnotation: metalk8s.scality.com/grafana-folder-name
```
Is the `-name` suffix needed? I mean, do we expect some other annotations regarding Grafana folders?

jbertran replied:
We don't, it just seems better to me to explain what impact this annotation has on the folder. I can change it if you'd like.

bert-e commented:
# History mismatch



Merge commit #96cb4e6849a0e93e274d727a76d583af3501cb9e on the integration branch
`w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces` is merging a branch which is neither the current
branch `improvement/3618-pick-up-dashboards-from-all-namespaces` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `improvement/3618-pick-up-dashboards-from-all-namespaces` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jbertran commented on salt/metalk8s/addons/prometheus-operator/deployed/chart.sls:
```diff
@@ -59893,7 +59896,7 @@ spec:
         unless ignoring(key,value) kube_node_spec_taint{job="kube-state-metrics",key=~"ToBeDeletedByClusterAutoscaler|cloud.google.com/impending-node-termination|aws-node-termination-handler/spot-itn"})
         == 1
-      for: 5m
+      for: 15m
```
Note: I can't figure out where this comes from - it doesn't look like that's from my diff, perhaps a value change that occurred without regeneration?

lavigne958 replied:
yes it is part of your commit, perhaps a small typo ??
https://github.com/scality/metalk8s/commit/88030820f89ecbdeeefce932b0ad12142c479d54

NicolasT replied:
I'll take a look.

NicolasT replied:
Looks like re-rendering the chart requires some 'special care': see the commit message of https://github.com/scality/metalk8s/commit/1c88876d2100baf04c898232f30531ab62f8f3c8

jbertran commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/3618-pick-up-dashboards-from-all-namespaces` into
`development/2.11`
* [w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast

NicolasT reviewed (dismissed):
Chart needs to be re-rendered with the 'right' command (see other comment).

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

bert-e commented:
# History mismatch



Merge commit #88030820f89ecbdeeefce932b0ad12142c479d54 on the integration branch
`w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces` is merging a branch which is neither the current
branch `improvement/3618-pick-up-dashboards-from-all-namespaces` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `improvement/3618-pick-up-dashboards-from-all-namespaces` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on charts/kube-prometheus-stack/charts/grafana/values.yaml:
```diff
@@ -630,12 +630,12 @@ sidecar:
     # If specified, the sidecar will search for dashboard config-maps inside this namespace.
```
Any reason to edit the grafana helm value file and not the `charts/kube-prometheus-stack.yaml` helm option file directly ?

NicolasT replied:
Heh, good point, indeed one isn't supposed to make changes to the vendored files unless really necessary.

jbertran replied:
Oh. That's just my weak knowledge of Helm. I saw `Values`, so I edited `Values` :sweat_smile:

NicolasT reviewed (commented):
Couple of other things:
- Needs CHANGELOG entry
- Likely needs some small documentation in the dev guide
- May benefit from a test?

Commit:
salt,charts: Grafana sidecar picks up dashboards in all namespaces

Configure k8s-sidecar in Grafana deployment:

* to search in `ALL` namespaces
* to place dashboards in folders according to
  `metalk8s.scality.com/grafana-folder-name` annotation
* to create Grafana folders from dashboard directory structure

Re-render chart to salt state using

```
./charts/render.py prometheus-operator \
  charts/kube-prometheus-stack.yaml \
  charts/kube-prometheus-stack/ \
  --namespace metalk8s-monitoring \
  --service-config grafana \
  metalk8s-grafana-config \
  metalk8s/addons/prometheus-operator/config/grafana.yaml \
  metalk8s-monitoring \
  --service-config prometheus \
  metalk8s-prometheus-config \
  metalk8s/addons/prometheus-operator/config/prometheus.yaml \
  metalk8s-monitoring \
  --service-config alertmanager \
  metalk8s-alertmanager-config \
  metalk8s/addons/prometheus-operator/config/alertmanager.yaml \
  metalk8s-monitoring \
  --service-config dex \
  metalk8s-dex-config \
  metalk8s/addons/dex/config/dex.yaml.j2 metalk8s-auth \
  --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
  --patch 'PrometheusRule,metalk8s-monitoring,prometheus-operator-kubernetes-system-kubelet,spec:groups:0:rules:1:for,"5m"' \
  > salt/metalk8s/addons/prometheus-operator/deployed/chart.sls
```

Fixes: #3618

jbertran commented:
> Couple of other things:
> 
> * Needs CHANGELOG entry
> 
> * Likely needs some small documentation in the dev guide
> 
> * May benefit from a test?

Changelog, docs update: done.
Tests: since the effect of this change is local to a generated `chart.sls` where the diff does not depend on Jinja to be rendered, I don't know what to test here.

gdemonet commented:
@jbertran For the test, I guess we'd build something E2E, creating a ConfigMap in some namespace, and check that Grafana reflects these changes. We can see how to implement it together, if you want.

jbertran commented:
> @jbertran For the test, I guess we'd build something E2E, creating a ConfigMap in some namespace, and check that Grafana reflects these changes. We can see how to implement it together, if you want.

It feels a bit like we're testing that various third party components work as advertised (basically `k8s-sidecar` provided directory structure and Grafana picking it up), but I can do E2E for sure - it's pytest so I'm not too lost there. I'll set up a bunch of namespaces and configmaps, push them and verify dashboards _and_ folders present. Does that sound sufficient?

NicolasT commented:
> It feels a bit like we're testing that various third party components work as advertised (basically k8s-sidecar provided directory structure and Grafana picking it up)

Well, yes indeed. If this functionality, no matter how it's provided, is part of the expected functionality of the project, it should be tested (granted, there's tests lacking in many places).
It's not really about "testing third parties", rather ensuring that, e.g., if upstream changes how they do things, and we update the chart(s), then we detect this change and must make relevant changes in MetalK8s to cope with it, ideally continuing to expose the feature as it is/was, otherwise changing the feature (kinda breaking change).

Charts and `values.yaml` have been... unstable APIs in the past, so we really shouldn't rely on upstream not 'breaking' things (sometimes involuntary).

TeddyAndrieux commented on CHANGELOG.md:
```diff
@@ -69,6 +69,11 @@
+- [#3618](https://github.com/scality/metalk8s/issues/3618) Detect Grafana
+  dashboard ConfigMaps in any namespace rather than just `metalk8s-monitoring`,
+  and enable Grafana folder generation from dashboard file structure (PR
+  [#3620](https://github.com/scality/metalk8s/pull/3620/commits))
```
```suggestion
  [#3620](https://github.com/scality/metalk8s/pull/3620))
```

TeddyAndrieux commented on tests/post/steps/test_monitoring.py:
```diff
@@ -119,10 +129,169 @@ def _check_object_exists():
+@given(parsers.parse("the '{ns_name}' namespace exists"))
+def namespace_exists(k8s_client, ns_name):
+    client = k8s_client.resources.get(api_version="v1", kind="Namespace")
+    namespaces = client.get().items
+    try:
+        next(
+            filter(lambda ns: ns.get("metadata", {}).get("name") == ns_name, namespaces)
+        )
+    except StopIteration:
+        try:
+            client.create({"metadata": {"name": ns_name}})
+        except ApiException as err:
+            raise AssertionError(f"Unable to create namespace: {err}")
```
Why iterate on all namespaces ?
You can retrieve only one (and create it if it does not exist)
Something like that
```suggestion
    try:
      client.get(name=ns_name)
    except ApiException as err:
      if err.status == 404:
        client.create({"metadata": {"name": ns_name}})
```

TeddyAndrieux commented on tests/post/steps/test_monitoring.py:
```diff
@@ -119,10 +129,169 @@ def _check_object_exists():
+            raise AssertionError(f"Unable to create namespace: {err}")
+
+
+@given(parsers.parse("the '{dash_name}' dashboard exists in folder '{folder_name}'"))
+def dashboard_exists(grafana_api, dash_name, folder_name):
+    folders = grafana_api.get_folders()
+    try:
+        folder = next(filter(lambda folder: folder["title"] == folder_name, folders))
+    except StopIteration:
+        raise AssertionError(f"no folder named '{folder_name}'")
+    dashboards = grafana_api.get_dashboards(folder_id=folder["id"])
+    try:
+        next(filter(lambda dash: dash["title"] == dash_name, dashboards))
+    except StopIteration:
+        raise AssertionError(
+            f"found no dashboard named '{dash_name}' in '{folder_name}'"
+        )
+    else:
+        return True
```
Maybe to make it simpler we could simply use the uid directly since we can set this one manually in the dashboard.

To make this test really "realistic" we could generate an `uuid` at the beginning of the test (put it in the context) and then we create a dashboard using this `uuid` as uid

TeddyAndrieux commented on tests/post/features/monitoring.feature:
```diff
@@ -56,3 +56,16 @@ Feature: Monitoring is up and running
+        When we put a dashboard as ConfigMap named 'zk-dash' in namespace 'zenko' with folder annotation value 'ZENKO-PLEASE'
+        Then we have a dashboard named 'zk-dash' in folder 'ZENKO-PLEASE'
+
+    Scenario: Removing a ConfigMap removes its dasboard
+        Given the Grafana API is available
+        And the 'zenko' namespace exists
+        And the 'zk-dash' dashboard exists in folder 'ZENKO-PLEASE'
+        When we delete the 'zk-dash' ConfigMap in namespace 'zenko'
+        Then we have no dashboard named 'zk-dash' in folder 'ZENKO-PLEASE'
```
Personally do not really like the fact that this test need the previous one to succeed.

To me we could have a single test with 2 "When - Then" (but maybe it's not a good practice I don't know)
Something like that 
```
Scenario: Inserting a new ConfigMap provides a new dashboard
    Given the Grafana API is available
    And the 'zenko' namespace exists
    When we put a dashboard as ConfigMap named 'zk-dash' in namespace 'zenko' with folder annotation value 'ZENKO-PLEASE'
    Then we have a dashboard named 'zk-dash' in folder 'ZENKO-PLEASE'
    When we delete the 'zk-dash' ConfigMap in namespace 'zenko'
    Then we have no dashboard named 'zk-dash' in folder 'ZENKO-PLEASE'
```

jbertran replied:
How about I switch this in the second test?
```         
- And the 'zk-dash' dashboard exists in folder 'ZENKO-PLEASE'
+ And the 'zk-dash' ConfigMap exists and deploys the 'zk-dash' dashboards in folder 'ZENKO-PLEASE'
```
This way we can deploy the ConfigMap if it's not available.

The goal of the second one is to make the first one's insertion of a strictly test-related ConfigMap (maybe there's an easy-to-use cleanup hook? I didn't find it :slightly_frowning_face: )

TeddyAndrieux replied:
But with the suggestion, I send to you the CM will be deleted at the end of the test also.

If you just want some clean up at the end of the test it's doable with some teardown  (you can check `tests/post/steps/test_ingress.py` we do some simple teardown)

TeddyAndrieux reviewed (dismissed):
LGTM just comment on test :smile:

Commit:
monitoring,tests: add E2E test for Grafana sidecar behavior

Grafana is expected to:

* Add dashboards as we add ConfigMaps with the right label to any namespace, in
  folders according to the `metalk8s.scality.com/grafana-folder-name` annotation
* Delete dashboards if the corresponding ConfigMaps are deleted.

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -119,10 +134,152 @@ def _check_object_exists():
+    try:
+        next(
+            filter(lambda cm: cm.get("metadata", {}).get("name") == cm_name, configmaps)
+        )
```
Why do we get the whole list if we only care about one?

jbertran replied:
fixed in `c60bea50e`

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -119,10 +134,152 @@ def _check_object_exists():
+    client = k8s_client.resources.get(api_version="v1", kind="Namespace")
+    try:
+        client.get(name=ns_name)
+    except ApiException:
```
This may need more specificity: we should only attempt to POST if GET returns a 404. Multiple issues may happen otherwise, such as an attempt to create after the GET returned a temporary 500, which will fail even if the API is back in a good shape.

gdemonet replied:
Also, could be nice to wrap these in retries, just in case... Well I guess we'll do it if we see the tests failing.

jbertran replied:
fixed in `c60bea50e` - leaving the retries out for now

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -119,10 +134,152 @@ def _check_object_exists():
+        try:
+            client._create(cm_manifest)
+        except ApiException as err:
+            raise AssertionError(f"Unable to push ConfigMap: {err}")
```
Why do we use a `finally:` block here? If the API is in a bad shape for some reason and refuses the deletion (or even the retrieval, which isn't wrapped in `try/except` block for now), we'll still try to create the ConfigMap.

jbertran replied:
Got lost in the control flow :sweat: fixed in `c60bea50e`

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -119,10 +134,152 @@ def _check_object_exists():
+
+
+@then(parsers.parse("the dashboard is no longer in folder '{folder_name}' in Grafana"))
+def check_no_dash_in_folder(context, gf_folder_name, grafana_api):
```
Should be:

```suggestion
@then(parsers.parse("the dashboard is no longer in folder '{gf_folder_name}' in Grafana"))
def check_no_dash_in_folder(context, gf_folder_name, grafana_api):
```

I imagine?

jbertran replied:
Oops, fixed in `c60bea50e`

gdemonet reviewed (changes requested):
A few things to change w.r.t. API error-handling, but LGTM otherwise! Thank you :+1:

gdemonet reviewed (commented):
LGTM :ok_hand: I'll approve once squashed!

jbertran commented:
@gdemonet done

gdemonet commented on tests/post/steps/test_monitoring.py:
```diff
@@ -148,6 +153,98 @@ def _check_object_exists():
+
+@given(parsers.parse("we have a dashboard ready with a specific UID"))
+def dashboard_ready(context):
+    cm_uid = str(uuid.uuid4())
```
Missing import for this module

jbertran replied:
:man_facepalming: conflict resolution gone wrong

jbertran commented:
/approve

bert-e commented:
# History mismatch



Merge commit #88030820f89ecbdeeefce932b0ad12142c479d54 on the integration branch
`w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces` is merging a branch which is neither the current
branch `improvement/3618-pick-up-dashboards-from-all-namespaces` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `improvement/3618-pick-up-dashboards-from-all-namespaces` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

jbertran commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/3618-pick-up-dashboards-from-all-namespaces` into
`development/2.11`
* [w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/3618-pick-up-dashboards-from-all-namespaces` and the
destination branch `development/2.11`.

Please resolve the conflict on **the feature branch** (`improvement/3618-pick-up-dashboards-from-all-namespaces`).

```sh
git fetch && \
git checkout origin/improvement/3618-pick-up-dashboards-from-all-namespaces && \
git merge origin/development/2.11
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/3618-pick-up-dashboards-from-all-namespaces
```







*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #55b586e7361183f4ad9ea02b1a207f7af7fafcad on the integration branch
`w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces` is merging a branch which is neither the current
branch `improvement/3618-pick-up-dashboards-from-all-namespaces` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `improvement/3618-pick-up-dashboards-from-all-namespaces` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

jbertran commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/3618-pick-up-dashboards-from-all-namespaces` into
`development/2.11`
* [w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/3618-pick-up-dashboards-from-all-namespaces) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/123.0`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/123.0`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye jbertran.
